### PR TITLE
Flatten structs for validation errors and warnings

### PR DESCRIPTION
**Pull request by john-h-kastner-aws**
## Description of changes

Building on #905 to finish #745 for validation errors and warnings. Flatten error and warnings structs, removing the `*Kind` enums. Also remove the `TypeError` struct entirely (`ValidationError` is now used everywhere).

Sorry for the massive diff. I did my best to split off some smaller PRs, but this has a large impact across a lot of validator tests.

Reviewers can focus their effort on the `cedar-policy` crate. With this change, the validation error types in `cedar-policy-validator` aren't publicly exported, so the public crates fully defines the new error API. Inside `cedar-policy-validator`, most of the changes are to tests, which I think you can mostly trust. I did need to make some changes in `typecheck.rs` which should be reviewed for correctness.

## Checklist for requesting a review

The change in this PR is (choose one, and delete the other options):

- [x] A breaking change requiring a major version bump to `cedar-policy` (e.g., changes to the signature of an existing API).

I confirm that this PR (choose one, and delete the other options):

- [x] Does not update the CHANGELOG because my change does not significantly impact released code.

I confirm that [`cedar-spec`](https://github.com/cedar-policy/cedar-spec) (choose one, and delete the other options):

- [x] Requires updates, and I have made / will make these updates myself. (Please include in your description a timeline or link to the relevant PR in `cedar-spec`, and how you have tested that your updates are correct.)

### CODE REVIEW

File: cedar-policy-validator/src/rbac.rs [hunk 1]
```diff
@@ -485,11 +499,9 @@ mod test {
...
             src,
             &Report::new(notes.first().unwrap().clone()),
-            &ExpectedErrorMessageBuilder::error(
-                "for policy `policy0`, unrecognized entity type `foo_type`",
-            )
-            .exactly_one_underline("foo_type")
-            .build(),
+            &ExpectedErrorMessageBuilder::error("unrecognized entity type `foo_type`")
+                .exactly_one_underline("foo_type")
+                .build(),
```
Comment (cdisselkoen) [inline cedar-policy-validator/src/rbac.rs:504]: is this a regression?  Similarly for other tests in this file

File: cedar-policy/src/api/err.rs [hunk 2]
```diff
@@ -434,82 +434,249 @@ impl From<cedar_policy_validator::SchemaError> for SchemaError {
     }
 }
 
+/// This module contains definitions for structs containing detailed information
+/// about each validation error. Errors are primarily documented on their
+/// variants in [ValidationError].
```
Comment (cdisselkoen) [inline cedar-policy/src/api/err.rs:439]:
```suggestion
/// variants in [`ValidationError`].
```

File: cedar-policy/src/api/err/validation_errors.rs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+#[cfg(doc)]
+use crate::ValidationError;
+
+// Generates a public struct wrapping a core validator error struct. The core
+// and external struct will have exactly the same name. This name _must_ be the
+// same as the name of the corresponding `ValidationError` variant in `err.rs`
+// (`cargo doc` will fail otherwise). This macro generate a basic doc-string
+// linking back to the `ValidationError` enum variant where the primary
+// documentation should be written.
+macro_rules! wrap_core_error {
```
Comment (cdisselkoen) [inline cedar-policy/src/api/err/validation_errors.rs:35]: Nice

**Review by cdisselkoen — APPROVED** [review]

**Comment by john-h-kastner-aws** [comment]
Should be good to go. Spec build failure fixed in cedar-policy/cedar-spec#340

**Review by shaobo-he-aws — APPROVED** [review]

**Review by cdisselkoen — APPROVED** [review]

**Pull request merged**